src/mbgl/mtl/drawable.cpp

```diff
@@ -189,8 +189,11 @@ void Drawable::draw(PaintParameters& parameters) const {
     encoder->setCullMode(cullMode.enabled ? mapCullMode(cullMode.side) : MTL::CullModeNone);
     encoder->setFrontFacingWinding(mapWindingMode(cullMode.winding));
 
-    if (auto state = shaderMTL.getRenderPipelineState(renderPassDescriptor, impl->vertexDesc, getColorMode())) {
-        encoder->setRenderPipelineState(state.get());
+    if (!pipelineState) {
+        pipelineState = shaderMTL.getRenderPipelineState(renderPassDescriptor, impl->vertexDesc, getColorMode());
+    }
```

Comment:
We need to recreate the pipelineState also if the `renderPassDescriptor`, `vertexDesc` or `getColorMode()` got changed.

Reply:
I think instead caching the RenderPipelineState on drawable, a better place will be caching in ShaderProgram.

Reply:
I'd agree with Tim here that Drawable or somewhere near that class is probably the right place.  Too many things to check in the program.  We can also just clear it out in Drawable if we change something.

I'll also add that we put it in Drawable in WhirlyGlobe and it never came up as an issue after that.

Reply:
We will still have to account for the render pass descriptor for clear colors, and if a drawable is allowed to be rendered to multiple targets.

Reply:
Ok, we should cache it on the drawable. The remaining issue is that the vertex descriptor is changing to often because of the vertex attributes being dirty. Not sure if this is a bug, because it force us to recreate the render pipeline state to often.

Reply:
In terms of what they intend for Metal, we're way off, but that's something we'll need to address incrementally by moving more work into the shaders and such.  In the short term, maybe we can make this work better.

Technically, we should only need to change the vertex descriptor object if the actual structure of the vertices changes.  That is, if we switch from a uniform to a vertex array, or something like that.  If we just put new data into one of the buffers for a vertex array, we don't need to update the vertex descriptor.

Reply:
This is will a fix, to check if the structure changes. But the underlaying issue remains, is it normal that the vertex attributes change so often? Maybe this can be another issue that decrease the performance.

Reply:
Right, that is the central issue.  The vertex attributes shouldn't change quite so often for our use case.  We can move most of that into the shaders.

But!  There are cases where the vertex attributes change constantly and we can make this path faster in Metal.  We just need to hold on to the vertex descriptors and update the buffers, rather than changing them each time.

Reply:
They talk about triple buffering in the Metal docs.  I'm not a fan of that, personally.  I prefer to use a blit encoder instead.

You set up your new data for the existing buffers, by creating new buffers.  Then you copy from the new buffers to the old ones.  If you combine that with a MTLHeap it's cheap enough.

Reply:
Yeah, we need to differentiate between the structure and the content being different.  Adding an attribute would make the structure dirty, while the bucket changing the shared data just requires re-uploading that buffer.

That latter flag is maintained in `VertexBufferBase`/`IndexBufferBase`, but the `VertexAttribute`/`VertexAttributeArray` level doesn't track the former.

I think we can assume that the structure of the attribute blocks won't change for a drawable, the layer should produce a new one in that case, but we should at least check that with an assertion.